packages/Kokkos/src/details/DTK_DetailsTreeConstruction.cpp

```diff
@@ -0,0 +1,343 @@
+    int step = upperBoundLenght;
+    do
+    {
+        step = step >> 1;
+        int newSplit = idx + ( split + step ) * direction;
+        //    unsigned int splitCode = sortedMortonCodes[newSplit];
+        if ( commonPrefix( sortedMortonCodes, numObjects, idx, newSplit ) >
+             commonPrefixLowerBound )
+            split += step;
+    } while ( step > 1 );
+    //  std::cout<<idx<<"  dir="<<direction<<"
+    //  delta_min="<<commonPrefixLowerBound<<"  lmax="<<upperBoundLenght<<"
+    //  length="<<split<<"\n";
+    int jdx = idx + split * direction;
+    // an equivalent to std::make_pair or std::minmax would be nice here
+    return Kokkos::pair<int, int>( min( idx, jdx ), max( idx, jdx ) );
+}
+
+// alternate top-down implementation less efficient
+Node *generateHierarchy( unsigned int *sortedMortonCodes, int *sortedObjectIDs,
+                         int first, int last )
+{
+    // Single object => create a leaf node.
+
+    if ( first == last )
+        return new LeafNode( &sortedObjectIDs[first] );
+
+    // Determine where to split the range.
+
+    int split = findSplit( sortedMortonCodes, first, last );
+
+    // Process the resulting sub-ranges recursively.
+
+    Node *childA =
+        generateHierarchy( sortedMortonCodes, sortedObjectIDs, first, split );
+    Node *childB = generateHierarchy( sortedMortonCodes, sortedObjectIDs,
+                                      split + 1, last );
+    return new InternalNode( childA, childB );
+}
+
+// to assign the Morton code for a given object, we use the centroid point of
+// its bounding box, and express it relative to the bounding box of the scene.
+void assignMortonCodes( AABB const *boundingBoxes, unsigned int *mortonCodes,
+                        int n, AABB const &sceneBoundingBox )
+{
+    std::array<double, 3> xyz;
+    double a, b;
+    for ( int i = 0; i < n; ++i ) // parallel for
+    {
+        for ( int d = 0; d < 3; ++d )
+        {
+            xyz[d] = 0.5 * ( boundingBoxes[i]._minmax[2 * d + 0] +
+                             boundingBoxes[i]._minmax[2 * d + 1] );
+            a = sceneBoundingBox._minmax[2 * d + 0];
+            b = sceneBoundingBox._minmax[2 * d + 1];
+            xyz[d] = ( xyz[d] - a ) / ( b - a );
+        }
+        mortonCodes[i] = morton3D( xyz[0], xyz[1], xyz[2] );
+    }
+}
+
+void sortObjects( unsigned int *morton_codes, int *object_ids, int n )
+{
+    using std::sort;
+    // possibly use thrust::sort()
+    // see https://thrust.github.io
+    sort( object_ids, object_ids + n,
+          [morton_codes]( int const &i, int const &j ) {
+              return morton_codes[i] < morton_codes[j];
+          } );
+    // TODO: in-place permutation of mortonCodes rather than 2nd sort
+    sort( morton_codes, morton_codes + n );
+}
+
+void calculateBoundingBoxOfTheScene( AABB const *boundingBoxes, int n,
+                                     AABB &sceneBoundingBox )
+{
+    // QUESTION: precondition on sceneBoundingBox?
```

Comment:
I do not reset the box so it is the user responsibility to give an uninitialized box.